### PR TITLE
Groundwork to be able to split partitions

### PR DESCRIPTION
**Pull request by harshagelivi**

### CODE REVIEW

File: api/protobuf-spec/worker.proto [hunk 1]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+package dstk;
+
+message SplitPartReq {
+  int64 id = 1;
+  repeated bytes splits = 2;
```
Comment (anujga) [inline api/protobuf-spec/worker.proto:6]: untyped?

File: cmd/disk-cache/app.go [hunk 2]
```diff
@@ -61,32 +26,31 @@ func mainRunner(conf string, cleanDb bool) error {
...
+		dcServer := MakeServer(ws.MsgHandler, chanSize)
+		dstk.RegisterDcRpcServer(ws.Server, dcServer)
+		return ws.Start("tcp", viper.GetString("url"))
 	})
-	err = f.Wait()
-	if err != nil {
-		return err
-	}
-	return nil
+	return f.Wait()
```
Comment (anujga) [inline cmd/disk-cache/app.go:53]: where are you shutting down WorkerServer

File: pkg/ss/handler.go [hunk 3]
```diff
@@ -6,11 +6,11 @@ import (
 )
 
 type MsgHandler struct {
-	WorkerActor
+	w WorkerActor
 }
 
 func (mh *MsgHandler) Handle(req Msg) ([]interface{}, error) {
-	mh.Mailbox() <- req
+	mh.w.Mailbox() <- req
```
Comment (anujga) [inline pkg/ss/handler.go]: error handling. write to channel can fail.

Comment (anujga) [reply pkg/ss/handler.go]: see if you can integrate some linter to always catch this error

File: pkg/ss/part.go [hunk 4]
```diff
@@ -46,24 +49,54 @@ func (p *PartRange) Id() int64 {
...
 func (p *PartRange) becomeRunningHandler() error {
 	p.smState = Running
+	var followerMailbox chan<- interface{}
 	for m := range p.mailBox {
-		p.consumer.Process(m)
+		switch m.(type) {
+		case ClientMsg:
+			cm := m.(ClientMsg)
+			p.consumer.Process(cm)
+			if !m.(ClientMsg).ReadOnly() && followerMailbox != nil {
```
Comment (anujga) [inline pkg/ss/part.go:58]: when is `followerMailbox == nil`

File: pkg/ss/part.go [hunk 5]
```diff
@@ -46,24 +49,54 @@ func (p *PartRange) Id() int64 {
...
+				followerMailbox <- cm
+			}
+		case *FollowRequest:
+			fr := m.(*FollowRequest)
+			followerMailbox = fr.followerMailbox
+			followerMailbox <- p.consumer.GetSnapshot()
+		default:
+			p.logger.Info("not handled", zap.Any("state", p.smState), zap.Any("type", reflect.TypeOf(m)))
```
Comment (anujga) [inline pkg/ss/part.go]: warning

**Review by anujga — APPROVED** [review]
Can u add some unit tests for the state machine

File: cmd/disk-cache/app.go (hunk 2 above)
Comment (harshagelivi) [reply cmd/disk-cache/app.go:53]: f.Wait() is a blocking call right? i didn't change any functionality, just cleaned up statements

File: api/protobuf-spec/worker.proto (hunk 1 above)
Comment (harshagelivi) [reply api/protobuf-spec/worker.proto:6]: i didn't understand. i wanted [][]byte, basically list of points in the range where we need to split.

File: pkg/ss/handler.go (hunk 3 above)
Comment (harshagelivi) [reply pkg/ss/handler.go]: added handling here, will check the custom linter part. doesn't seem to straightforward, will spend time on this after split is complete.

File: pkg/ss/part.go (hunk 4 above)
Comment (harshagelivi) [reply pkg/ss/part.go:58]: followerMailbox is set only when FollowRequest is received. zero value of channel is nil right?

File: pkg/ss/part.go (hunk 5 above)
Comment (harshagelivi) [reply pkg/ss/part.go]: done

**Comment by harshagelivi** [comment]
> Can u add some unit tests for the state machine

will add after split is completed. this and custom lint.

File: api/protobuf-spec/worker.proto (hunk 1 above)
Comment (anujga) [reply api/protobuf-spec/worker.proto:6]: my mistake; `type key = []bytes`

**Pull request merged**